keps/sig-testing/20190118-breaking-apart-the-kubernetes-test-tarball.md

```diff
@@ -0,0 +1,274 @@
+    - [Risks and Mitigations](#risks-and-mitigations)
+    - [Test Plan](#test-plan)
+    - [Graduation Criteria](#graduation-criteria)
+  - [References](#references)
+  - [Implementation History](#implementation-history)
+
+## Release Signoff Checklist
+
+- [x] [k/enhancements issue in release milestone and linked to KEP](https://github.com/kubernetes/enhancements/issues/714)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+The Kubernetes release artifacts include a "mondo" test tarball which includes both
+"portable" test sources (such as shell scripts and image manifests) as well as
+platform-specific test binaries for all supported client, node, and
+server platforms.
+
+This KEP proposes replacing this monolith test tarball with many platform-specific
+tarballs, matching the existing pattern used for the client, node, and server
+tarballs.
+
+## Motivation
+
+As the number of supported client, server, and node platforms has increased, the
+size of the test tarball has increased as well. As of Kubernetes v1.13.2, the
+official `kubernetes-test.tar.gz` is approximately 1.2GB; previous releases
+ranged from 1.3 - 1.5GB. Several years ago,
+[there were complaints](https://github.com/kubernetes/kubernetes/issues/28435)
+that the "full" `kubernetes.tar.gz` release tarball was too big at 1.4G.
+Many of the motivations for breaking up that tarball echo into this proposal.
+
+Another motivation comes from the Bazel effort. It's possible to build all
+release artifacts solely using Bazel, and there is progress being made on
+supporting cross-compilation of binary artifacts. There is no support for
+combining artifacts for multiple platforms in one Bazel invocation, however, so
+it is not possible to produce a monolith test tarball using Bazel. Separating
+the test tarball would thus help with the "Bazel parity" effort, though it is
+not the sole motivation.
+
+### Goals
+
+- The Kubernetes test artifacts are broken apart such that users only need to
+  download binaries for platforms they're using.
+- Be largely invisible to most developers: everything should just keep working.
+
+### Non-Goals
+
+- Changing the underlying build system. Both the make/shell-based build
+  system and the Bazel-based build system will be affected, but users can
+  continue to use their existing workflow.
+- Removing cruft from the test tarballs. Likely, there are binaries and portable
+  sources no longer being used anywhere, but we won't prune them with this effort.
+- Changing what is released outside of the test tarball.
```

Comment:
Hi @ixdy,

Perhaps call out `kubernetes.tar.gz` and its `/cluster` directory as an example here...

Reply:
that's actually not quite what I was referring to. At the sig-testing meeting some folks mentioned wanting to be able to download things like `e2e.test` separately, which I don't want to address here.

Reply:
in particular, this was basically a direct response to https://github.com/kubernetes/enhancements/pull/701#discussion_r249637475.